### PR TITLE
Feat: SqueezeOp

### PR DESCRIPTION
**Pull request by vasantha-helprack**
# Added 
- SqueezeOp:
  - V11
  - V13
- Lit tests

### CODE REVIEW

**Review by ayonam-helprack — DISMISSED** [review]
Generally, the changes made for formatting the code should be placed in a separate PR.

Adding them to a Feature PR makes it a little difficult to separate the grain from the chaff and review the changes made for the feature.  :-)

File: src/Conversion/ONNXToTorch/NN/CommonUtils.cpp [hunk 1]
```diff
@@ -92,8 +97,43 @@ std::vector<Value> createArrayAttribute(::mlir::ArrayAttr onnxArrayAttr,
...
+Value getIntValue(int val, ConversionPatternRewriter &rewriter,
+                  mlir::MLIRContext *context, Location loc) {
+  auto iType = IntegerType::get(context, 64);
+  auto iVal = IntegerAttr::get(iType, val);
+  return rewriter.create<ConstantIntOp>(loc, iVal);
+}
+
+std::vector<int> toVector(mlir::ArrayAttr axesAttr) {
+  std::vector<int> axes;
+
```
[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/CommonUtils.cpp]: Done

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+// Copyright 2022, Helprack LLC.
+//
+// =============================================================================
+//
+// This file lowers most unary operators from torch to onnx using a template
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToTorch/NN/CommonUtils.h"
+#include "src/Conversion/ONNXToTorch/ONNXToTorchCommon.hpp"
```
[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp:15]: Done

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+  for (auto x : axesSet) {
+    // positive integers are added as it
+    // negative integers are normarlized to positive
+    axesNonNeg.push_back((x > 0) ? x : (x + axesSet.size()));
+  }
+  return axesNonNeg;
+}
+
+std::vector<int> getSortedWithNegativeAxes(mlir::ArrayAttr axesRaw) {
```
[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp]: Done

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------- SqrtOp.cpp - ONNX Op Transform ------------------===//
```
Comment (BraunPhilipp) [inline src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp:5]: nit: rename please

Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp]: Done

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+std::vector<int> getAxes(ONNXSqueezeOp squeezeOp) {
+  auto builder = mlir::Builder(squeezeOp.getContext());
+  auto axesConstOp = getONNXConstantOp(squeezeOp.axes());
+  auto axesAttr = createArrayAttrFromConstantOp(builder, axesConstOp);
+  return getSortedWithNonNegativeAxes(axesAttr);
+}
+
+std::vector<int> getAxes(ONNXSqueezeV11Op squeezeOp) {
```
Comment (BraunPhilipp) [inline src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp:56]: Please add some explanation regarding differences between `ONNXSqueezeV11Op` and `ONNXSqueezeOp`. When did you incur them? Is it related to different op sets?

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+#include <set>
+#include <vector>
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+std::vector<int> toUniqueAndNonNegative(std::vector<int> axes) {
+  std::set<int> axesSet(axes.begin(), axes.end());
+  std::vector<int> axesNonNeg;
```
Comment (BraunPhilipp) [inline src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp:29]: `std::set` is sorted by default. You can probably remove the reordering lateron.

**Review by BraunPhilipp — CHANGES_REQUESTED** [review]

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp (hunk 6 above)
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp]: Done

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp (hunk 5 above)
Comment (BraunPhilipp) [reply src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp:56]: Thanks for looking into it!

File: test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop11_unsorted_axis.onnx.mlir [hunk 7]
```diff
@@ -0,0 +1,13 @@
+//RUN: onnx-mlir --EmitONNXIR --run-torch-pass %s -o=%t >/dev/null && cat %t.onnx.mlir | FileCheck -v %s
+module attributes {}  {
+  func @main_graph(%arg0: tensor<4x1x4x3x1x4x1xf32>) -> tensor<4x4x3x4xf32> attributes {input_names = ["input"], output_names = ["output"]} {
+//CHECK: %[[CONST1:.*]] = torch.constant.int 1
+//CHECK: [[RES1:%.]] = torch.aten.squeeze.dim %arg0, %[[CONST1]] : !torch.vtensor<[4,1,4,3,1,4,1],f32>, !torch.int -> !torch.vtensor<[4,4,3,4],f32>
```
Comment (BraunPhilipp) [inline test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop11_unsorted_axis.onnx.mlir:5]: This is functionally not correct. `aten.squeeze` is only applied to the dimension you are specifying.

The result type needs to change here. Maybe look at flatten as a reference.

File: test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop11_non-negative_axis.onnx.mlir [hunk 8]
```diff
@@ -0,0 +1,9 @@
+//RUN: onnx-mlir --EmitONNXIR --run-torch-pass %s -o=%t >/dev/null && cat %t.onnx.mlir | FileCheck -v %s
```
Comment (BraunPhilipp) [inline test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop11_non-negative_axis.onnx.mlir:1]: nit: Please rename `squeezeop` to `squeeze` in the `*.mlir` files.

Comment (BraunPhilipp) [inline test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop11_non-negative_axis.onnx.mlir:1]: nit: Please rename `squeezeop` to `squeeze` in the `*.mlir` files.

File: test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop13_non-negative_axis.onnx.mlir [hunk 9]
```diff
@@ -0,0 +1,10 @@
+//RUN: onnx-mlir --EmitONNXIR --run-torch-pass %s -o=%t >/dev/null && cat %t.onnx.mlir | FileCheck -v %s
+module attributes {}  {
+  func @main_graph(%arg0: tensor<?x1x2x1x2xf32>) -> tensor<?x2x1x2xf32> attributes {input_names = ["input"], output_names = ["output"]} {
+    %0 = "onnx.Constant"() {onnx_node_name = "Constant_0", value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
+//CHECK: %[[DIM:.*]] = torch.constant.int 1
+    %1 = "onnx.Squeeze"(%arg0, %0) {onnx_node_name = "Squeeze_1"} : (tensor<?x1x2x1x2xf32>, tensor<1xi64>) -> tensor<?x2x1x2xf32>
+//CHECK: torch.aten.squeeze.dim %arg0, %[[DIM]] : !torch.vtensor<[?,1,2,1,2],f32>, !torch.int -> !torch.vtensor<[?,2,1,2],f32>
```
Comment (BraunPhilipp) [inline test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop13_non-negative_axis.onnx.mlir:7]: This is also wrong. For `torch.constant.int 0` the result type would be correct. Below you can find a corresponding example I generated from `torch-mlir`.

```
module attributes {torch.debug_module_name = "Model"} {
  func @forward(%arg0: !torch.vtensor<[1,3,128,128],f32>) -> !torch.vtensor<[1,3,128,128],f32> {
    %int1 = torch.constant.int 1
    %0 = torch.aten.squeeze.dim %arg0, %int1 : !torch.vtensor<[1,3,128,128],f32>, !torch.int -> !torch.vtensor<[1,3,128,128],f32>
    return %0 : !torch.vtensor<[1,3,128,128],f32>
  }
}
```

Comment (BraunPhilipp) [inline test/mlir/conversion/onnx_to_torch/squeezeop/squeezeop13_non-negative_axis.onnx.mlir:7]: This is also wrong. For ``torch.constant.int 0` the result type would be correct.

```
module attributes {torch.debug_module_name = "Model"} {
  func @forward(%arg0: !torch.vtensor<[1,3,128,128],f32>) -> !torch.vtensor<[1,3,128,128],f32> {
    %int1 = torch.constant.int 1
    %0 = torch.aten.squeeze.dim %arg0, %int1 : !torch.vtensor<[1,3,128,128],f32>, !torch.int -> !torch.vtensor<[1,3,128,128],f32>
    return %0 : !torch.vtensor<[1,3,128,128],f32>
  }
}
```

**Review by BraunPhilipp — CHANGES_REQUESTED** [review]

File: src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    for (uint64_t i = 0; i < axes.size(); i++) {
+      // With every successive deleting on dimension, the input axis
+      // changes to `axis = axis - number_of_dimensions_deleted`
+      // This works because, axes is sorted and normalized to possitive integers
+      auto dim_raw = axes[i] - i;
+      // Fail when dimension squeeze is not 1?
+      // assert((result.getType().dyn_cast<TensorType>().getShape()[dim_raw] ==
+      // 1) && "Cannot squeeze for dim");
+      Value dim = getIntValue(dim_raw, rewriter, context, loc);
+      result = rewriter.create<AtenSqueezeDimOp>(loc, resultType, result, dim);
```
[the remark this answers is not in the document]
Comment (vasantha-helprack) [reply src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp]: Done

[the remark this answers is not in the document]
Comment (BraunPhilipp) [reply src/Conversion/ONNXToTorch/NN/SqueezeOp.cpp]: Thanks!

**Review by BraunPhilipp — APPROVED** [review]

**Review by siboehm — APPROVED** [review]
This seems to have been one of the most complex operators. LGTM!

**Review by ljfitz — APPROVED** [review]

**Pull request merged**